### DISCUSSION TITLE
CQL ThreadPoolExecutor has a fixed size

### DISCUSSION THREAD

**Discussion by cdegroc** [JanusGraph/janusgraph#2545] [Ideas]
Hi all,

`CQLStoreManager/CQLKeyColumnValueStore` instantiate an `ExecutorService` that is used to convert [CQL ResultSets to a JanusGraph EntryList](https://github.com/JanusGraph/janusgraph/blob/aa3964207b107844de9ce0917f1e2a74843a13ac/janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java#L281-L290), among other things.

That ExecutorService is declared [here (master branch)](https://github.com/JanusGraph/janusgraph/blob/aa3964207b107844de9ce0917f1e2a74843a13ac/janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLStoreManager.java#L179-L187) as follows
```
this.executorService = new ThreadPoolExecutor(
    10,                             // corePoolSize
    100,                            // maximumPoolSize
    1,                              // keepAliveTime
    TimeUnit.MINUTES,               // unit
    new LinkedBlockingQueue<>(),    // workQueue
    new ThreadFactoryBuilder()      // threadFactory
            .setDaemon(true)
            .setNameFormat("CQLStoreManager[%02d]")
            .build());
```

Given that corePoolSize=10 and maximumPoolSize=100, I was expecting that ThreadPool to dynamically grow/shrink as needed. However instrumentation (using an [InstrumentedExecutorService](https://github.com/dropwizard/metrics/blob/release/4.1.x/metrics-core/src/main/java/com/codahale/metrics/InstrumentedExecutorService.java)) showed that the ThreadPool was always using 10 threads.

This is explained in the ThreadPoolExecutor [javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ThreadPoolExecutor.html):
> Using an unbounded queue (for example a LinkedBlockingQueue without a predefined capacity) will cause new tasks to wait in the queue when all corePoolSize threads are busy. Thus, no more than corePoolSize threads will ever be created. (And the value of the maximumPoolSize therefore doesn't have any effect.)

I think we could fix this in two ways:
1. Make the code clearer by switching to a `newFixedThreadPool(10, new ThreadFactoryBuilder()...)`.
2. Make the ThreadPool really dynamic by using a custom ThreadPoolExecutor implementation (see [this article](https://dzone.com/articles/scalable-java-thread-pool-executor) for more details and [this ElasticSearch implementation](https://github.com/elastic/elasticsearch/blob/dfac67aff0ca126901d72ed7fe862a1e7adb19b0/server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java#L74-L81) for reference).

Happy to get feedback and contribute a PR if there's interest.
I'm also curious to know if this could be a bottleneck in JanusGraph (and also, if not, why).

**Comment by porunov** [comment]
Thank you for starting this topic @cdegroc !
I would prefer a second option. I.e. I would imagine the pool to act something like cached thread pool which respects `maximumPoolSize` and `corePoolSize`. I think it would be great to see a contribution in this area. 
I also think that `corePoolSize` and `maximumPoolSize` should be configurable via properties.

**Comment by li-boxuan** [comment]
Actually I don't even understand why we need a thread pool here. 😅  Can someone explain? Thanks!

**Reply by cdegroc** [reply]
The thread pool was introduced in [this commit](https://github.com/JanusGraph/janusgraph/commit/2aac19f69b190fb3338189405d264ab32b98877a).

Here is my understanding:

Before this change the following construct was used in [CQLKeyColumnValueStore.getSlice](https://github.com/JanusGraph/janusgraph/blob/d928f5f5fe7a9f3e1acd619e96492325a99363b4/janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java#L266-L272)
```
Future<EntryList> result = Future
  .fromJavaFuture(this.session.executeAsync(this.getSlice.bind()...)
  .map(resultSet -> fromResultSet(resultSet, this.getter));
interruptibleWait(result);
```
which means `fromResultSet(resultSet, this.getter)` was scheduled to [run on vavr DEFAULT_EXECUTOR](https://github.com/vavr-io/vavr/blob/v0.10.3/vavr/src/main/java/io/vavr/concurrent/Future.java#L283-L286) which is [ForkJoinPool.commonPool()](https://github.com/vavr-io/vavr/blob/v0.10.3/vavr/src/main/java/io/vavr/concurrent/Future.java#L85-L111) and the await would be waiting on that pool's thread.

By introducing a new thread pool, we have more control over that thread.

**Reply by li-boxuan** [reply]
Thanks for the detailed explanation! I guess my confusion actually comes from: why do we need to use vavr Future at the first place? Maybe I am missing something, but why not just
```
Future future = this.session.executeAsync(this.getSlice.bind().....
return fromResultSet(future.get(), this.getter);
```

**Reply by cdegroc** [reply]
It was there from the start. Your guess is as good as mine :)
Could another JanusGraph dev have more insights?
Otherwise I can test without it and observe the effects on threads states and performance metrics but that will require more time.

**Reply by li-boxuan** [reply]
Ping the authors: @pkendall64 @samantmaharaj

**Reply by li-boxuan** [reply]
I am just guessing: maybe the author(s) wanted to use the thread pool as a bulkhead to limit the number of concurrent backend requests?

**Reply by li-boxuan** [reply]
FYI @porunov created a PR: https://github.com/JanusGraph/janusgraph/pull/2700/files to make thread pool size configurable

**Comment by porunov** [comment]
@li-boxuan , @cdegroc I am thinking to change the default configuration for CQL ExecutorService to be disabled by default after #2741 . I'm not changing the default configuration in that PR to not debate about the default configurations in the PR but if that PR is merged I was thinking to propose to change the default settings to be disabled and add additional information to the documentation so that users would know that this option is available historically but isn't recommended. What do you think about it?  

Performance tests conducted:
https://github.com/JanusGraph/janusgraph/pull/2741#issuecomment-894331436
https://github.com/JanusGraph/janusgraph/pull/2741#issuecomment-894635159
https://github.com/JanusGraph/janusgraph/pull/2741#issuecomment-894656248
https://github.com/JanusGraph/janusgraph/pull/2741#issuecomment-894663680

**Reply by li-boxuan** [reply]
I don't have a strong opinion but I am a bit inclined to defer this to the next (major) release. It might be the case that with the ExecutorService disabled, heavy workloads will start to fail in production environments, and users will have to tune the `max-requests-per-connection`. [This](https://github.com/JanusGraph/janusgraph/issues/2406#issuecomment-884014382) is an example where the workload is so heavy that queries start to fail. This is less likely to happen when the executor service is used because it limits the number of concurrent requests.

That being said, I am fine with releasing it in this release if we **emphasize** that existing users should be careful and keep ExecutorService enabled before they thoroughly test and tune the configs.

**Reply by porunov** [reply]
Good thoughts about tuning. Maybe we should leave executor service enabled for this release but emphasize that's its recommended to disable it in production.

**Comment by porunov** [comment]
CQL ThreadPoolExecutor was extended to support multiple different implementations (as well as custom implementations). Moreover, CQL storage backend doesn't use ThreadPoolExecutor to manage IO operations anymore. Instead, it reuses DataStax Async approach and uses ThreadPoolExecutor for deserialization work only. Thus, fixed thread pool of cores * 2 is now preferable size in most cases.
See PR #3760 for more information. Closing this discussion as resolved, but feel free to reopen in case we missed anything.
